File: src/one-stroke/one-stroke.cc
#include <one-stroke/one-stroke.h>

namespace puzzle_solver {
  OneStroke::OneStroke() {}

  std::vector<std::vector<int>> OneStroke::solve(std::vector<std::vector<std::pair<int, int>>>& graph) {
    std::vector<std::vector<int>> ans;
    
  }

  std::vector<std::vector<std::pair<int, int>>> OneStroke::createGraph(const std::vector<std::vector<int>>& matrix) {
    std::vector<std::vector<std::pair<int, int>>> graph;
    for (size_t i = 0; i < matrix.size(); i++) {
      graph.push_back(std::vector<std::pair<int,int>>());
      for (size_t j = 0; j < matrix[0].size(); j++) {
        if (matrix[i][j] > 0) {
          graph[i].push_back(std::pair<int,int>(j, matrix[i][j]));
        }
      }
    }

    return graph;
  }
}

File: puzzles/one_stroke.py
class OneStroke:
  def __init__(self):
    pass

  def create_graph(self, matrix):
    graph = []
    for i, row in enumerate(matrix):
      graph.append([])
      for j, elem in enumerate(row):
        if elem > 0:
          graph[i].append((j, elem))
    return graph

  def solve(self, graph):
    ans = []
    pass

  def __solvable(self, graph, curr, ans):
    if self.__valid_solve(graph):
      return True

    for edge in graph[curr]:
      if edge[1] > 0:
        edge[1] -= 1
        ans.append((curr, edge[0]))
        if self.__solvable(graph, edge[0], ans):
          return True
        edge[1] +=1
        ans.pop()
    return False

  def __valid_solve(self, graph):
    for node in graph:
      for neighbor in node:
        if neighbor[1] > 0:
          return False
    return True